### ISSUE TITLE
Sampling performance improvement

### ISSUE THREAD

**Issue by dbalabka** [cgevans/scikits-bootstrap#29]
I've noticed that bootstrap sampling time increases not linearly with increasing dataset size which leads to a significant performance decrease. As result, I've found that Numpy array sampling can be improved by using [recommended Numpy Generator API](https://numpy.org/doc/stable/reference/random/generator.html) can be used.

The current approach of sampling based on old `numpy.random.randint`:
```python
        bootindices = bootstrap_indices(tdata[0], n_samples)
        stat = np.array([statfunction(*(x[indices] for x in tdata))
                            for indices in bootindices])
```
![image](https://user-images.githubusercontent.com/1269372/115530827-cd6cf580-a29c-11eb-96ad-5f208af39e9c.png)

Refactored sampling using [recommended Numpy Generator API](https://numpy.org/doc/stable/reference/random/generator.html) which improves performance for large datasets:
```python
        rng = np.random.default_rng()
        stat = np.array([statfunction(*(rng.choice(x, tdata[0].shape[0]) for x in tdata)) for _ in range(n_samples)])
```
![image](https://user-images.githubusercontent.com/1269372/115530692-aca4a000-a29c-11eb-8c73-f58989a57fac.png)

Using Numpy Generator API instead of our own sampling implementation we can improve performance drastically.

**Note!** It solves only half of the problem. Another issue is sampling based on Jackknife Indexes:
```python
        jackindices = jackknife_indices(tdata[0])
        jstat = [statfunction(*(x[indices] for x in tdata))
                for indices in jackindices]
````

**Comment by dbalabka** [comment]
Current, Jackknife resampling performance improvement:
```python
        # jackknife_indices creates multiple copies of array by using np.delete
        jackindices = jackknife_indices(tdata[0])
        jstat = [statfunction(*(x[indices] for x in tdata))
                for indices in jackindices]
```
![image](https://user-images.githubusercontent.com/1269372/115439086-dadf9c80-a216-11eb-9de6-6c34db0c72a1.png)

can be improved as follows:
```python
        # keeping only single array mask
        index_bool = np.ones((len(tdata[0]),), bool)
        jstat = []
        for i in np.arange(0, len(tdata[0])):
            # skip single row from array
            index_bool[i] = False
            jstat.append(statfunction(*(x[index_bool] for x in data)))
            # restore
            index_bool[i] = True
```
![image](https://user-images.githubusercontent.com/1269372/115439308-1b3f1a80-a217-11eb-9630-d4ee6c1794db.png)

**Comment by cgevans** [comment]
Thanks - I'll try to take a look at these this evening.  Looking at this briefly on my phone, a concern might be that it inlines the sampling, and people might want to use the _indices functions separately. But the speedups seem pretty significant.


On 20 April 2021 18:32:36 IST, Dmitry Balabka ***@***.***> wrote:
>Current, Jackknife resampling performance improvement:
>```python
>        # jackknife_indices creates multiple copies of array by using np.delete
>        jackindices = jackknife_indices(tdata[0])
>        jstat = [statfunction(*(x[indices] for x in tdata))
>                for indices in jackindices]
>```
>![image](https://user-images.githubusercontent.com/1269372/115439086-dadf9c80-a216-11eb-9de6-6c34db0c72a1.png)
>
>can be improved as follows:
>```python
>        # keeping only single array mask
>        index_bool = np.ones((len(tdata[0]),), bool)
>        jstat = []
>        for i in np.arange(0, len(tdata[0])):
>            # skip single row from array
>            index_bool[i] = False
>            jstat.append(statfunction(*(x[index_bool] for x in data)))
>            # restore
>            index_bool[i] = True
>```
>![image](https://user-images.githubusercontent.com/1269372/115439308-1b3f1a80-a217-11eb-9630-d4ee6c1794db.png)
>
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/cgevans/scikits-bootstrap/issues/29#issuecomment-823468837
-- 
Sent from my phone. Please excuse my brevity.

**Comment by dbalabka** [comment]
@cgevans thanks that take a look! IMO `*_indices` can stay for backward compatibility reasons.

**Comment by dbalabka** [comment]
@cgevans sorry, but I've found a bug in my bootstrapping implementation. Now, it is only an x2 performance boost which sounds more realistic. I've updated the description with a new stat. This is still a nice achievement.

**Comment by dbalabka** [comment]
@cgevans earlier in #28 I've mentioned that the library can be optimized using Numba JIT. After bottlenecks have been identified we can easily optimize them using Numba JIT. The major benefit of Numba in this case that we can have backwards-compatible code and there no need to install Numba. The main limitation of the Numba in the context of the application for this library is that Numba does not support callbacks. So we do not have the possibility to pass any custom stat function as an argument. Below proposal focused mostly on default library behaviour when we use the un-weighted `np.average` function which is equivalent to `np.mean`. IMO it still worth optimizing the library for this specific case because IMO is the most widely used. For example, AB testing in the case of e-Commerce requires to calculate confidence intervals for average metrics such as CVR, AOV, Average Revenue per Session:
https://medium.com/criteo-engineering/why-your-ab-test-needs-confidence-intervals-bec9fe18db41

# Boostrapping resampling for `np.mean`
Unfortunately, my initial proposal had a bug and performance improvement isn't good enough. So I focused on the Numba version with `np.mean`:
```python
@numba.njit(parallel=True, fastmath=True)
def _calculate_boostrap_mean_stat(data: np.ndarray, n_samples: int) -> np.ndarray:
    n = data.shape[0]
    stat = np.zeros(n_samples)
    for i in numba.prange(n_samples):
        stat[i] = np.random.choice(data, n).mean()
    return stat
```

As result, I've achieved significant performance improvement which can be increased using more CPU cores. The code has been tested on 4 cores. Here is the testing script:
https://gist.github.com/dbalabka/9c0a3d88312f6d51af8949e85758aa7a

# Jackknife resampling for `np.mean`
I've implemented calculation logic in case if we want to use `np.mean`. It became super fast because we do not need to calculate the mean each time when we remove a single element from the dataset. Even non compiled version takes less than a second to calculate the Jackknife statistics. Numba's version is slower than pure Python implementation because it takes about one second to compile. Still, it became faster than the pure Python version when the dataset contains more then 10*10^6 rows in dataset.

```python
@numba.njit(parallel=True, fastmath=True)
def _calculate_jackknife_mean_stat(data: np.ndarray) -> np.ndarray:
    n = data.shape[0]
    jstat = np.zeros(n)
    sum = data.sum()
    for i in numba.prange(n):
        # Alternative solution, which can be used when we want use custom stat function
        # jstat[i] = np.concatenate((data[:i], data[i+1:])).mean()
        jstat[i] = (sum - data[i]) / (n - 1)
    return jstat
```

Here is the testing script with result:
https://gist.github.com/dbalabka/68fb2cdd6353a549273b56f57515637b

**Comment by dbalabka** [comment]
Here is a prototype how it can be integrated:
https://github.com/cgevans/scikits-bootstrap/pull/30

**Comment by cgevans** [comment]
Implemented in 1.1.0.
